src/main/java/cam72cam/immersiverailroading/model/TrackModel.java

```diff
@@ -1,45 +1,298 @@
 package cam72cam.immersiverailroading.model;
 
+import cam72cam.immersiverailroading.library.TrackModelPart;
+import cam72cam.immersiverailroading.track.BuilderBase;
+import cam72cam.immersiverailroading.track.BuilderBase.VecYawPitch;
+import cam72cam.immersiverailroading.util.DataBlock;
+import cam72cam.immersiverailroading.util.MathUtil;
+import cam72cam.immersiverailroading.util.RailInfo;
 import cam72cam.mod.model.obj.OBJModel;
+import cam72cam.mod.render.obj.OBJRender;
+import cam72cam.mod.render.opengl.VBO;
 import cam72cam.mod.resource.Identifier;
 import trackapi.lib.Gauges;
+import util.Matrix4;
 
-import java.util.ArrayList;
+import java.util.*;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
 
-public class TrackModel extends OBJModel {
+public class TrackModel extends OBJModel{
+    private TrackOrder order;
+    //Get randomized idents' real ident
+    private Map<String, Supplier<String>> randomMap;
+    //Map primitive idents to OBJModel
+    private final Map<String, Map<TrackModelPart, List<String>>> groupNamesMapper;
     private final String compare;
     private final double size;
-    private final double height;
+    private double height;
     public final double spacing;
 
-    public TrackModel(String condition, Identifier resource, double model_gauge_m, double spacing) throws Exception {
-        super(resource, 0, Gauges.STANDARD / model_gauge_m);
+    public TrackModel(String condition, Identifier resource, double modelGaugeM, double spacing, boolean isSingle) throws Exception {
+        super(resource, 0, Gauges.STANDARD / modelGaugeM);
         this.compare = condition.substring(0, 1);
-        this.size = Double.parseDouble(condition.substring(1));
-        ArrayList<String> groups = new ArrayList<>();
-        for (String group : groups()) {
-            if (group.contains("RAIL_LEFT") || group.contains("RAIL_RIGHT")) {
-                groups.add(group);
+        this.groupNamesMapper = new HashMap<>();
+        if(isSingle) {
+            Map<TrackModelPart, List<String>> groups = new HashMap<>();
+            for(TrackModelPart part : TrackModelPart.values()){
+                List<String> parts = this.groups().stream().filter(part::is).collect(Collectors.toList());
+                groups.put(part, parts);
             }
+            groupNamesMapper.put("single", groups);
+            randomMap.put("single", () -> "single");
```

Comment:
Is this necessary?  Shouldn't the order look up the group directly?

Reply:
Ah, you treat random as the main lookup and add singleton entries below as well.  Clever :+1: